include/ncengine/ecs/ComponentPool.h

```diff
@@ -0,0 +1,301 @@
+{
+    public:
+        using value_type = T;
+        using iterator = std::vector<T>::iterator;
+        using const_iterator = std::vector<T>::const_iterator;
+        using reverse_iterator = std::vector<T>::reverse_iterator;
+
+        /** @brief Construct a new component pool. */
+        explicit ComponentPool(size_t maxEntities, ComponentHandler<T> handler)
+            : m_storage{maxEntities}, m_handler{std::move(handler)} {}
+
+        /** @brief Emplace a component attached to an entity. */
+        template<class... Args>
+        auto Add(Entity entity, Args&&... args) -> T*;
+
+        /** @brief Get a pointer to the component attached to an entity, or nullptr if one does not exist. */
+        auto Get(Entity entity) -> T*;
+
+        /** @brief Get a const pointer to the component attached to an entity, or nullptr if one does not exist. */
+        auto Get(Entity entity) const -> const T*;
+
+        /** @brief Get the entity a component is attached to, or a null entity on failure. */
+        auto GetParent(const T* component) const -> Entity;
+
+        /** @brief View all components as a contiguous range. */
+        auto GetComponents() noexcept -> std::span<T> { return m_storage.GetPackedArray(); }
+
+        /** @brief View all components as a contiguous range. */
+        auto GetComponents() const noexcept -> std::span<const T> { return m_storage.GetPackedArray(); }
+
+        /** @brief Get the T's ComponentHandler. */
+        auto Handler() noexcept -> ComponentHandler<T>& { return m_handler; }
+
+        /** @brief Get the T's onAdd event Signal. */
+        auto OnAdd() noexcept -> Signal<T&>& { return m_onAdd; }
+
+        /** @brief Get the T's onRemove event Signal. */
+        auto OnRemove() noexcept -> Signal<Entity>& { return m_onRemove; }
+
+        /** @brief Sort the components according to a predicate. */
+        template<std::predicate<const T&, const T&> Predicate>
+        void Sort(Predicate&& compare);
+
+        /** @brief Get an iterator the the first component in the pool. */
+        auto begin() noexcept { return std::ranges::begin(m_storage.GetPackedArray()); }
+
+        /** @brief Get a const_iterator the the first component in the pool. */
+        auto begin() const noexcept { return std::ranges::begin(m_storage.GetPackedArray()); }
+
+        /** @brief Get an iterator one past the last component in the pool. */
+        auto end() noexcept { return std::ranges::end(m_storage.GetPackedArray()); }
+
+        /** @brief Get a const_iterator one past the last component in the pool. */
+        auto end() const noexcept { return std::ranges::end(m_storage.GetPackedArray()); }
+
+        /** @brief Get the number of components committed to the pool. */
+        auto size() const noexcept { return Size(); }
+
+        /** @brief Check if there are no components in the pool. */
+        [[nodiscard]] auto empty() const noexcept { return GetComponents().empty(); }
```

Comment:
I made the signatures for these match exactly what is in the stl. `empty()` now has the `[[nodiscard]]` attribute simply because its a frequently misinterpreted name, and the attribute guarantees to catch the misunderstanding. (clear the container vs. check if its empty).